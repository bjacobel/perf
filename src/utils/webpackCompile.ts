import {
  Stats,
  Compiler,
  Configuration,
  ConfigurationFactory,
  MultiConfigurationFactory,
} from 'webpack';
import path from 'path';
import merge from 'webpack-merge';

// polymorphism is evil
type AmbiguousConfig =
  | Configuration
  | Configuration[]
  | ConfigurationFactory
  | MultiConfigurationFactory;

// webpack.config.js can have a terrifying array of export types. This normalizes them all to an array.
const getConfig = async (ambi: AmbiguousConfig): Promise<Configuration[]> => {
  const appConfigs: Configuration[] = [];

  if (typeof ambi === 'object') {
    if (ambi.hasOwnProperty('length')) {
      return Promise.resolve(ambi as Configuration[]);
    } else {
      appConfigs.push(ambi as Configuration);
    }
  } else if (typeof ambi === 'function') {
    const factoryResult = await ambi(
      { production: true },
      { mode: 'production' },
    );
    appConfigs.push(...(await getConfig(factoryResult)));
  }

  return appConfigs;
};

export default async (...extraConfig: Configuration[]) => {
  const cwd = process.cwd();

  const appConfigs = await getConfig(
    require(path.join(cwd, 'webpack.config.js')),
  );

  const finalConfigs = appConfigs.map((config) =>
    merge(config, ...extraConfig, {
      optimization: {
        noEmitOnErrors: true,
      },
    }),
  );

  const webpack: (c: Configuration[]) => Compiler = require(path.join(
    cwd,
    'node_modules/webpack',
  ));

  const compiler = webpack(finalConfigs);
  return new Promise((resolve, reject) => {
    compiler.run((err: Error, stats: Stats) => {
      if (err || stats.hasErrors()) {
        reject(err || new Error(stats.toJson().errors.join('\n')));
      } else {
        resolve(stats.toJson().assets);
      }
    });
  });
};
