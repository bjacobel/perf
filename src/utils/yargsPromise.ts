import yargs, { Argv, CommandModule } from 'yargs';

export default async (cmd: CommandModule, args: string = '') => {
  return new Promise((resolve, reject) =>
    yargs
      .command(cmd)
      .fail(reject)
      .parse(args, (err: Error, _argv: Argv, output: string) => {
        if (err) {
          reject(err);
        } else {
          resolve(output);
        }
      }),
  );
};
