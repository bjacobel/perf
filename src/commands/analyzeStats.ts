import { Arguments, CommandBuilder } from 'yargs';
import os from 'os';
import path from 'path';
import { promises } from 'fs';
import { Stats } from 'webpack';
const fs = promises;

import webpackCompile from '../utils/webpackCompile';
import {
  prettyKb,
  prettyDiff,
  renderTable,
  TableEntry,
} from '../utils/sizeDiff';

export const command = 'analyze-stats [compareBaseline] [emitBaseline]';

export const describe =
  'display sizes reported by compilation stats file, optionally compared to a previous run';

export const builder: CommandBuilder = {
  compareBaseline: {
    type: 'boolean',
  },
  emitBaseline: {
    type: 'boolean',
  },
};

const tableEntry = (
  stats: Stats.ToJsonOutput,
  baseline: Stats.ToJsonOutput | undefined,
) => {
  const matchingBaseline = (name: string): { size: number } | false => {
    const bail = false;
    if (!baseline) return bail;
    if (!baseline.assets) return bail;
    return baseline.assets.find((asset) => asset.name === name) ?? bail;
  };
  return (stats.assets ?? [])
    .filter((asset) => !/\.map$/.test(asset.name))
    .map(({ name, size }) =>
      Object.assign(
        {
          name,
          size: prettyKb(size),
        },
        matchingBaseline(name)
          ? {
              diff: prettyDiff(
                size,
                (matchingBaseline(name) as { size: number }).size,
              ),
            }
          : {},
      ),
    );
};

const extractSizes = (
  stats: Stats.ToJsonOutput,
  baseline: Stats.ToJsonOutput | undefined,
): TableEntry[] => {
  const childSizes: TableEntry[] = (stats.children ?? [])
    .map((child, i) => extractSizes(child, baseline && baseline.children![i]))
    .reduce((a, b) => [...a, ...b], []);

  return [...tableEntry(stats, baseline), ...childSizes];
};

export const handler = async (args: Arguments) => {
  let baseline: Stats.ToJsonOutput | undefined;
  const baselinePath = path.join(os.tmpdir(), 'baseline.json');

  if (args.compareBaseline) {
    baseline = JSON.parse(await fs.readFile(baselinePath, 'utf-8'));
  }

  const stats = await webpackCompile({
    // Remove hashes from filenames, even though we're building in prodmode
    // @TODO: CSS files are still hashed, because that's set inside MiniCssExtractPlugin instantiation
    output: {
      filename: '[name].js',
    },
  });

  renderTable(!!baseline, extractSizes(stats, baseline));

  if (args.emitBaseline) {
    await fs.writeFile(baselinePath, JSON.stringify(stats));
    console.log(
      `Emitted baseline stats file. Run next time with --compare-baseline to diff against this point in time.`,
    );
  }
};
