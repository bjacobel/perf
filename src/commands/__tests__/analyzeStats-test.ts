import { Stats } from 'webpack';
import { promises } from 'fs';

import * as analyzeStats from '../analyzeStats';
import webpackCompile from '../../utils/webpackCompile';
import yargsPromise from '../../utils/yargsPromise';

jest.mock('../../utils/webpackCompile');

const assets = (
  ...sizes: { [filename: string]: number }[]
): { name: string; size: number }[] =>
  sizes.map((size) => ({
    name: Object.keys(size)[0],
    size: Object.values(size)[0],
  }));

describe('analyze-stats command', () => {
  let fakeStats: Stats.ToJsonOutput;

  beforeEach(() => {
    jest.spyOn(promises, 'readFile');
    jest.spyOn(promises, 'writeFile');
    console.log = jest.fn();
    fakeStats = {
      assets: assets({ top: 2000 }),
      children: [
        { assets: assets({ small: 200 }, { big: 4000 }, { medium: 600 }) },
      ],
    } as Stats.ToJsonOutput;
    jest.mocked(webpackCompile).mockReturnValueOnce(Promise.resolve(fakeStats));
  });

  it('runs a webpack build and reports a table of stats', async () => {
    await yargsPromise(analyzeStats, 'analyze-stats');

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(jest.mocked(console.log).mock.calls[0][0]).toMatchSnapshot();
  });

  it('writes a baseline if sent --emit-baseline', async () => {
    await yargsPromise(analyzeStats, 'analyze-stats --emit-baseline');

    expect(promises.writeFile).lastCalledWith(
      expect.any(String),
      JSON.stringify(fakeStats),
    );
  });

  it('runs a webpack build and reports a table of stats compared to a baseline', async () => {
    const baseline = {
      assets: assets({ top: 7000 }),
      children: [{ assets: assets({ big: 4000 }, { medium: 205 }) }],
    } as Stats.ToJsonOutput;
    jest
      .mocked(promises.readFile)
      .mockReturnValueOnce(Promise.resolve(JSON.stringify(baseline)));

    await yargsPromise(analyzeStats, 'analyze-stats --compare-baseline');

    expect(jest.mocked(console.log).mock.calls[0][0]).toMatchSnapshot();
  });
});
