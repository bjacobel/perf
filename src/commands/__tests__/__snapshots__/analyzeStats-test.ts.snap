// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze-stats command runs a webpack build and reports a table of stats 1`] = `
"┌────────┬─────────┐
│ name   │    size │
├────────┼─────────┤
│ top    │ 1.95 kB │
├────────┼─────────┤
│ small  │   200 B │
├────────┼─────────┤
│ big    │ 3.91 kB │
├────────┼─────────┤
│ medium │   600 B │
└────────┴─────────┘"
`;

exports[`analyze-stats command runs a webpack build and reports a table of stats compared to a baseline 1`] = `
"┌────────┬─────────┬───────────────┐
│ name   │    size │ baseline diff │
├────────┼─────────┼───────────────┤
│ top    │ 1.95 kB │      -4.88 kB │
├────────┼─────────┼───────────────┤
│ small  │   200 B │               │
├────────┼─────────┼───────────────┤
│ big    │ 3.91 kB │     unchanged │
├────────┼─────────┼───────────────┤
│ medium │   600 B │        +395 B │
└────────┴─────────┴───────────────┘"
`;
